test/compile-test.cc

```diff
@@ -139,13 +139,136 @@ TEST(CompileTest, FormatWideString) {
   EXPECT_EQ(L"42", fmt::format(FMT_COMPILE(L"{}"), 42));
 }
 
+struct test_custom_formattable {};
+
+FMT_BEGIN_NAMESPACE
+template <> struct formatter<test_custom_formattable> {
+  enum class output_type { two, four } type{output_type::two};
+
+  FMT_CONSTEXPR auto parse(format_parse_context& ctx) -> decltype(ctx.begin()) {
+    auto it = ctx.begin(), end = ctx.end();
+    while (it != end && *it != '}') {
+      ++it;
+    }
+    auto spec = string_view(ctx.begin(), static_cast<size_t>(it - ctx.begin()));
+    auto tag = string_view("custom");
+    if (spec.size() == tag.size()) {
+      bool is_same = true;
+      for (size_t index = 0; index < spec.size(); ++index) {
+        if (spec[index] != tag[index]) {
+          is_same = false;
+          break;
+        }
+      }
+      type = is_same ? output_type::four : output_type::two;
+    } else {
+      type = output_type::two;
+    }
+    return it;
+  }
+
+  template <typename FormatContext>
+  auto format(const test_custom_formattable&, FormatContext& ctx) const
+      -> decltype(ctx.out()) {
+    return format_to(ctx.out(), type == output_type::two ? "{:>2}" : "{:>4}",
+                     42);
+  }
+};
+FMT_END_NAMESPACE
+
 TEST(CompileTest, FormatSpecs) {
   EXPECT_EQ("42", fmt::format(FMT_COMPILE("{:x}"), 0x42));
+  EXPECT_EQ("42", fmt::format(FMT_COMPILE("{}"), test_custom_formattable()));
+  EXPECT_EQ("  42",
+            fmt::format(FMT_COMPILE("{:custom}"), test_custom_formattable()));
 }
 
-TEST(CompileTest, DynamicWidth) {
+struct test_dynamic_formattable {};
+
+FMT_BEGIN_NAMESPACE
+template <> struct formatter<test_dynamic_formattable> {
+  size_t amount = 0;
+  detail::arg_ref<char> width_refs[3];
+
+  FMT_CONSTEXPR auto parse(format_parse_context& ctx) -> decltype(ctx.begin()) {
+    amount = static_cast<size_t>(*ctx.begin() - '0');
+    if (amount >= 1) {
+      width_refs[0] = detail::arg_ref<char>(ctx.next_arg_id());
+    }
+    if (amount >= 2) {
+      width_refs[1] = detail::arg_ref<char>(ctx.next_arg_id());
+    }
+    if (amount >= 3) {
+      width_refs[2] = detail::arg_ref<char>(ctx.next_arg_id());
+    }
+    return ctx.begin() + 1;
+  }
+
+  template <typename FormatContext>
+  auto format(const test_dynamic_formattable&, FormatContext& ctx) const
+      -> decltype(ctx.out()) {
+    int widths[3]{};
+    for (size_t i = 0; i < amount; ++i) {
+      detail::handle_dynamic_spec<detail::width_checker>(widths[i],
+                                                         width_refs[i], ctx);
+    }
+    if (amount == 1) {
+      return format_to(ctx.out(), "{:{}}", 41, widths[0]);
+    } else if (amount == 2) {
+      return format_to(ctx.out(), "{:{}}{:{}}", 41, widths[0], 42, widths[1]);
+    } else if (amount == 3) {
+      return format_to(ctx.out(), "{:{}}{:{}}{:{}}", 41, widths[0], 42,
+                       widths[1], 43, widths[2]);
+    } else {
+      throw format_error("formatting error");
+    }
+  }
+};
+FMT_END_NAMESPACE
```

Comment:
Same here. duration formatter has dynamic field support.

Reply:
Actually, the previous one (about custom formatter) and this are not the same.
Yes, it has dynamic field support. But as far as I can see, it supports the same set of nested replacement fields as the default formatter, `{:{}.{}}`. So handling 2 dynamic fields for the default formatter would probably be enough to pass the test with `chrono::duration` formatter.
While this custom formatter has a custom syntax for nested replacement fields (non `{:{}.{}}`), and it has 3 of them. So handling default dynamic fields wouldn't be enough to pass the test.

Reply:
I don't think we need to test the implementation of exotic formatter specializations here.

Reply:
Done with format string from chrono-test that uses dynamic specs